File: src/main/java/org/trp/shincolle/menu/DeskMenu.java
package org.trp.shincolle.menu;

import net.minecraft.core.BlockPos;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.DataSlot;
import net.minecraft.world.item.ItemStack;
import org.trp.shincolle.block.entity.DeskBlockEntity;

public class DeskMenu extends AbstractContainerMenu {
    private final int deskType;
    private int chapter;
    private int page;
    private int guiFunc;
    private int radarZoom;
    private final DeskBlockEntity blockEntity;
    private final boolean clientSide;

    public DeskMenu(int id, Inventory playerInventory, RegistryFriendlyByteBuf data) {
        this(id, playerInventory, getDeskTypeAndEntity(playerInventory, data));
    }

    private static Object[] getDeskTypeAndEntity(Inventory playerInventory, RegistryFriendlyByteBuf data) {
        int deskType = data.readInt();
        if (deskType == 0) {
            BlockPos pos = data.readBlockPos();
            if (playerInventory.player.level().getBlockEntity(pos) instanceof DeskBlockEntity desk) {
                return new Object[]{deskType, desk};
            }
            throw new IllegalStateException("Desk block entity not found.");
        }
        return new Object[]{deskType, data.readInt(), data.readInt()};
    }

    private DeskMenu(int id, Inventory playerInventory, Object[] typeAndData) {
        this(id, playerInventory, (int) typeAndData[0], 
             typeAndData.length > 2 ? (int) typeAndData[1] : 0, 
             typeAndData.length > 2 ? (int) typeAndData[2] : 0,
             0, 0,
             typeAndData.length == 2 ? (DeskBlockEntity) typeAndData[1] : null);
    }

    public DeskMenu(int id, Inventory playerInventory, int deskType) {
        this(id, playerInventory, deskType, 0, 0, 0, 0, null);
    }

    public DeskMenu(int id, Inventory playerInventory, int deskType, int chapter, int page) {
        this(id, playerInventory, deskType, chapter, page, 0, 0, null);
    }

    public DeskMenu(int id, Inventory playerInventory, int deskType, int chapter, int page, int guiFunc, int radarZoom, DeskBlockEntity blockEntity) {
        super(ModMenus.DESK_MENU.get(), id);
        this.deskType = deskType;
        this.chapter = chapter;
        this.page = page;
        this.guiFunc = guiFunc;
        this.radarZoom = radarZoom;
        this.blockEntity = blockEntity;
        this.clientSide = playerInventory.player.level().isClientSide;

        if (this.blockEntity != null) {
            this.chapter = this.blockEntity.getBookChap();
            this.page = this.blockEntity.getBookPage();
            this.guiFunc = this.blockEntity.getGuiFunc();
            this.radarZoom = this.blockEntity.getRadarZoomLv();

            this.addDataSlot(new DataSlot() {
                @Override
                public int get() {
                    return DeskMenu.this.blockEntity.getBookChap();
                }

                @Override
                public void set(int value) {
                    DeskMenu.this.chapter = value;
                }
            });

            this.addDataSlot(new DataSlot() {
                @Override
                public int get() {
                    return DeskMenu.this.blockEntity.getBookPage();
                }

                @Override
                public void set(int value) {
                    DeskMenu.this.page = value;
                }
            });

            this.addDataSlot(new DataSlot() {
                @Override
                public int get() {
                    return DeskMenu.this.blockEntity.getGuiFunc();
                }

                @Override
                public void set(int value) {
                    DeskMenu.this.guiFunc = value;
                }
            });

            this.addDataSlot(new DataSlot() {
                @Override
                public int get() {
                    return DeskMenu.this.blockEntity.getRadarZoomLv();
                }

                @Override
                public void set(int value) {
                    DeskMenu.this.radarZoom = value;
                }
            });
        }
    }

    @Override
    public ItemStack quickMoveStack(Player player, int index) {
        return ItemStack.EMPTY;
    }

    @Override
    public boolean stillValid(Player player) {
        return true;
    }

    public int getDeskType() {
        return deskType;
    }

    public int getChapter() {
        return chapter;
    }

    public int getPage() {
        return page;
    }

    public int getGuiFunc() {
        return guiFunc;
    }

    public int getRadarZoom() {
        return radarZoom;
    }

    public DeskBlockEntity getBlockEntity() {
        return blockEntity;
    }
}

File: src/main/java/org/trp/shincolle/menu/DeskMenu.kt
package org.trp.shincolle.menu

import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.ContainerLevelAccess
import net.minecraft.world.inventory.DataSlot
import net.minecraft.world.item.ItemStack
import org.trp.shincolle.block.entity.DeskBlockEntity
import org.trp.shincolle.item.DeskItemBook
import org.trp.shincolle.item.DeskItemRadar
import org.trp.shincolle.server.TeamDiplomacyService.sendDeskDiplomacySync

class DeskMenu @JvmOverloads constructor(
    id: Int,
    playerInventory: Inventory,
    @JvmField val deskType: Int,
    @JvmField var chapter: Int = 0,
    @JvmField var page: Int = 0,
    @JvmField var guiFunc: Int = 0,
    @JvmField var radarZoom: Int = 0,
    val blockEntity: DeskBlockEntity? = null
) : AbstractContainerMenu(ModMenus.DESK_MENU.get(), id) {
    private val clientSide: Boolean

    constructor(id: Int, playerInventory: Inventory, data: RegistryFriendlyByteBuf) : this(
        id,
        playerInventory,
        getDeskTypeAndEntity(playerInventory, data)
    )

    private constructor(id: Int, playerInventory: Inventory, typeAndData: Array<Any?>) : this(
        id, playerInventory, typeAndData[0] as Int,
        if (typeAndData.size > 2) typeAndData[1] as Int else 0,
        if (typeAndData.size > 2) typeAndData[2] as Int else 0,
        0, 0,
        if (typeAndData.size == 2) typeAndData[1] as DeskBlockEntity? else null
    )

    init {
        this.clientSide = playerInventory.player.level().isClientSide

        if (this.blockEntity != null) {
            this.chapter = this.blockEntity.getBookChap()
            this.page = this.blockEntity.getBookPage()
            this.guiFunc = this.blockEntity.getGuiFunc()
            this.radarZoom = this.blockEntity.getRadarZoomLv()

            this.addDataSlot(object : DataSlot() {
                override fun get(): Int {
                    return this@DeskMenu.blockEntity.getBookChap()
                }

                override fun set(value: Int) {
                    this@DeskMenu.chapter = value
                }
            })

            this.addDataSlot(object : DataSlot() {
                override fun get(): Int {
                    return this@DeskMenu.blockEntity.getBookPage()
                }

                override fun set(value: Int) {
                    this@DeskMenu.page = value
                }
            })

            this.addDataSlot(object : DataSlot() {
                override fun get(): Int {
                    return this@DeskMenu.blockEntity.getGuiFunc()
                }

                override fun set(value: Int) {
                    this@DeskMenu.guiFunc = value
                }
            })

            this.addDataSlot(object : DataSlot() {
                override fun get(): Int {
                    return this@DeskMenu.blockEntity.getRadarZoomLv()
                }

                override fun set(value: Int) {
                    this@DeskMenu.radarZoom = value
                }
            })

            if (!this.clientSide && this.guiFunc >= 3 && this.guiFunc <= 4 && playerInventory.player is ServerPlayer) {
                sendDeskDiplomacySync(playerInventory.player as ServerPlayer)
            }
        }
    }

    override fun quickMoveStack(player: Player, index: Int): ItemStack {
        return ItemStack.EMPTY
    }

    override fun stillValid(player: Player): Boolean {
        if (blockEntity != null) {
            if (blockEntity.getLevel() == null || player.level()
                    .getBlockEntity(blockEntity.getBlockPos()) !== blockEntity
            ) {
                return false
            }
            return stillValid(
                ContainerLevelAccess.create(blockEntity.getLevel()!!, blockEntity.getBlockPos()),
                player,
                blockEntity.getBlockState().getBlock()
            )
        }
        if (player.isRemoved || !player.isAlive) {
            return false
        }
        return when (this.deskType) {
            1 -> player.getMainHandItem().getItem() is DeskItemRadar
                    || player.getOffhandItem().getItem() is DeskItemRadar

            2 -> player.getMainHandItem().getItem() is DeskItemBook
                    || player.getOffhandItem().getItem() is DeskItemBook

            else -> true
        }
    }

    companion object {
        private fun getDeskTypeAndEntity(playerInventory: Inventory, data: RegistryFriendlyByteBuf): Array<Any?> {
            checkNotNull(data) { "Missing desk menu data." }
            val deskType = data.readInt()
            if (deskType == 0) {
                val pos = data.readBlockPos()
                if (playerInventory.player.level().getBlockEntity(pos) is DeskBlockEntity) {
                    val desk = playerInventory.player.level().getBlockEntity(pos) as DeskBlockEntity
                    return arrayOf<Any?>(deskType, desk)
                }
                throw IllegalStateException("Desk block entity not found.")
            }
            return arrayOf<Any?>(deskType, data.readInt(), data.readInt())
        }
    }
}
